### PR TITLE
mongo_proxy: support configurable command list for metrics

### LINKED ISSUE #13448: mongo_proxy: support configurable command list for metrics
It'd be nice to be able to generate metrics for additional commands, beyond "insert", "update", and "delete" [0].  We could just add a bunch of new commands to this hardcoded list, but that has a few problems:
1) It's not future-proof against new MongoDB releases adding new commands.
2) Our MongoDB deployment contains some custom diagnostic or administrative commands that wouldn't make sense to upstream.

Assuming this seems reasonable, I'm happy to provide a patch for this functionality.

[0] https://github.com/envoyproxy/envoy/blob/master/source/extensions/filters/network/mongo_proxy/mongo_stats.cc#L31

### PR DESCRIPTION
**Pull request by bartle-stripe**
Commit Message: mongo_proxy: support configurable command list for metrics

Additional Description:
* This additionally removes `query` from the default list of commands, since that's not actually a MongoDB command.  I'm not sure why it was ever added, but we never see any `query` command metrics (not be to be confused with the separate `query` namespace).
* Hardcoding aliases (`findandmodify` => `findAndModify`, etc...) seems reasonable, and doesn't make sense to be configurable, because these aliases only exist for historical reasons, and newer MongoDB commands (afaik) don't have such aliases.

Risk Level: Low

Testing: Updated tests and tested internally at Stripe.

Docs Changes: Updated `mongo_proxy` config docs to mention new `commands` field.

Release Notes: mongo_proxy: the list of commands to produce metrics for is now configurable.

Fixes #13448

### CODE REVIEW

File: api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto [hunk 1]
```diff
@@ -19,6 +19,7 @@ option (udpa.annotations.file_status).package_version_status = FROZEN;
 // MongoDB :ref:`configuration overview <config_network_filters_mongo_proxy>`.
 // [#extension: envoy.filters.network.mongo_proxy]
 
+// [#next-free-field: 6]
```
Comment (mattklein123) [inline api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto:22]: v2 is locked/deprecated, so please remove these changes.

File: api/envoy/extensions/filters/network/mongo_proxy/v3/mongo_proxy.proto [hunk 2]
```diff
@@ -39,4 +40,9 @@ message MongoProxy {
   // Flag to specify whether :ref:`dynamic metadata
   // <config_network_filters_mongo_proxy_dynamic_metadata>` should be emitted. Defaults to false.
   bool emit_dynamic_metadata = 4;
+
+  // List of commands to emit metrics for. Defaults to "delete", "insert", and "update".
+  // Note that metrics will not be emitted for "find" commands, since those are considered
+  // queries, and metrics for those are emitted under a dedicated "query" namespace.
+  repeated string commands = 5;
```
Comment (mattklein123) [inline api/envoy/extensions/filters/network/mongo_proxy/v3/mongo_proxy.proto:47]: Please add a release note.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Looks great, thanks! Just a few small comments.

/wait

File: api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto (hunk 1 above)
Comment (bartle-stripe) [reply api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto:22]: Hm, so I originally didn't include v2 in this change but I got test failures locally.

File: api/envoy/extensions/filters/network/mongo_proxy/v3/mongo_proxy.proto (hunk 2 above)
Comment (bartle-stripe) [reply api/envoy/extensions/filters/network/mongo_proxy/v3/mongo_proxy.proto:47]: Done.

File: api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto (hunk 1 above)
Comment (bartle-stripe) [reply api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto:22]: Specifically, a segfault in what looked to be an attempt to convert v3 into v2 config.

Comment (bartle-stripe) [reply api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto]: Anyway, removed, hopefully CI passes.

Comment (bartle-stripe) [reply api/envoy/config/filter/network/mongo_proxy/v2/mongo_proxy.proto]: Looks good, I think I might have been testing this change internally on an older git commit.

**Comment by mattklein123** [comment]
(Friendly request to please never force push. It makes reviews much more difficult. Thank you!)

File: api/envoy/extensions/filters/network/mongo_proxy/v3/mongo_proxy.proto (hunk 2 above)
Comment (mattklein123) [reply api/envoy/extensions/filters/network/mongo_proxy/v3/mongo_proxy.proto:47]: I don't see a release note? Please also ref link to this field. You will need to merge main to pick up the CI change for docs also.

/wait

Comment (bartle-stripe) [reply api/envoy/extensions/filters/network/mongo_proxy/v3/mongo_proxy.proto:47]: Sorry, should be there now.

**Comment by mattklein123** [comment]
@bartle-stripe please stop force pushing! Thanks!

**Comment by bartle-stripe** [comment]
Sorry, the instructions on the DCO check told me to force push my last commit because I didn't include `Signed-off-by `.

File: source/extensions/filters/network/mongo_proxy/config.cc [hunk 3]
```diff
@@ -34,7 +34,14 @@ Network::FilterFactoryCb MongoProxyFilterConfigFactory::createFilterFactoryFromP
     fault_config = std::make_shared<Filters::Common::Fault::FaultDelayConfig>(proto_config.delay());
   }
 
-  auto stats = std::make_shared<MongoStats>(context.scope(), stat_prefix);
+  auto commands = std::vector<std::string>{"delete", "insert", "update"};
+  if (proto_config.commands_size() > 0) {
+    commands =
+        std::vector<std::string>(proto_config.commands().begin(), proto_config.commands().end());
+    ;
```
Comment (mattklein123) [inline source/extensions/filters/network/mongo_proxy/config.cc:41]: nit: errant ';'

**Review by mattklein123 — CHANGES_REQUESTED** [review]
LGTM with small nit, thanks!

/wait

File: source/extensions/filters/network/mongo_proxy/config.cc (hunk 3 above)
Comment (bartle-stripe) [reply source/extensions/filters/network/mongo_proxy/config.cc]: Good catch, fixed!

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Comment by mattklein123** [comment]
Apologies this will need another main merge.

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**